README.md
# synomin7
SPK's for DSM7 | Current and Latest

CREDITs:

toafez für das DSM7DemoSPK, das Toolset und der Kreativität: https://github.com/toafez/DSM7DemoSPK
vletroye für die Impressionen in Sachen CGI Router und der SSH Administrator Abfrage während der Installation
gnadelwartz für die eigentliche arbeit an Synomin
webmin für die tolle Verwaltungsoberfläche

Vorwort:

Ich übernehme keinerlei Garantien, Gewährleistung oder sonstiges bei beschädigungen oder unsachgemäßer verwendung. Das was hier angeboten wird richtet sich an erfahrene Nutzer die wissen was sie tun.

Ich möchte mich bei @Toafez aka Tommes (https://github.com/toafez) sowie @Vletroye aka Valéry (https://github.com/vletroye) bedanken die sich intensiv mit der DSM7 Problematik auseinandergesetzt haben und dieses Paket hier zumindest zu 95% zum Leben erwecken konnten =)

Ebenfalls ein herzliches dankeschön an @Gnadelwartz aka Kay (https://github.com/gnadelwartz) der das Ursprüngliche Synomin Package zusammengestellt hat.

Ich kann nun selber gut nachvollziehen was für ein Riesen Sackstand das dank Synology ist, nachdem ich über 10 Tage im Trial and Error Verfahren damit verbracht habe die Verknüpfung zur Synology Oberfläche wiederherzustellen.

Ebenfalls noch Danke an das Team von Webmin - eine geniale Oberfläche die keine Wünsche offen lässt.

Die angebotenen SPK Pakete beziehen sich dabei auf die install.sh Datei im Target/Scripts Ordner:

Current >> http://download.webmin.com/devel/tarballs/
Latest >>> http://builds.webmin.dev/

Bei der Latest fehlen die deutschen Sprachdateien, diese können dann über das Current Archiv nachträglich hochgeladen werden. ( webmin/lang & webmin/authentic-theme/lang )

Eines noch vorweg: es ist kein 100% automatisiertes Installationspaket - das war mir leider nicht möglich, ebenfalls scheinen Updates nicht zu funktionieren ( access denied error )
In dem Paket befinden sich vermutlich Dateien die nicht gebraucht werden, ich erhoffe mir von dem "Release" das es vielleicht jemand noch optimieren kann da ich derzeit etwas entnervt bin

Installation

Im Vorfeld sind Perl ( Wahlweise Synology oder *PKG ) sowie auch IPKG/OPKG zu installieren
Die empfohlenen Module für *PKG sind im Vorfeld zu installieren: git dstat lm-sensors per-net-ssleay
Die optionalen Module ebenfalls im Vorfeld ( wer Bedarf hat ): coreutils git git-manpages dstat man man-pages lm-sensors iptables mc openssl perl perl-net-ssleay lsof gcc make

Das Archiv etc enthält Dateien für die Root Access Geschichte - ob sie nun gebraucht werden kann jeder selber testen
Die Datei /var/packages/webmin/target/ui/app_permissions.sh hat bei mir immer einen Merkwürdigen Output in der Console zur Folge gebracht:

"title: 'webmin:app:title' is neither mail string key nor i18n format." - der User webmin ist dennoch danach teil der Admin Gruppe geworden.

Die entsprechende SPK eurer Wahl Mittels DSM Paketverwaltung hochladen, die Warnung ignorieren ( Thirdparty ), euer User steht schon in der SSH Abfrage, Passwort händisch nachtragen.

Die Installation läuft durch, Webmin stoppen und per SSH einloggen.

Den Pfad auf /var/packages/webmin/target/scripts/ in Putty wechseln und die ./install.sh ausführen.

Im Anschluss habe ich noch händisch im Terminal Benutzer und Gruppe geändert ( Pfad: /var/packages/webmin/target/webmin/:

sudo chown -R webmin:webmin /webmin/

Das wars auch schon


Änderungen / Selber kompilieren:

Der Inhalt des working Archivs muss in ./working/* entpackt werden so das folgende Struktur entsteht:

/working/etc
/working/man
/working/reset
/working/scripts
/working/update

Bitte der Anleitung von Toafez folgen:

https://github.com/toafez/DSM7DemoSPK

Lizenz

Dieses Programm ist freie Software. Sie können es unter den Bedingungen der GNU General Public License, wie von der Free Software Foundation veröffentlicht, weitergeben und/oder modifizieren, entweder gemäß Version 3 der Lizenz oder (nach Ihrer Option) jeder späteren Version.

Die Veröffentlichung dieses Programms erfolgt in der Hoffnung, daß es Ihnen von Nutzen sein wird, aber OHNE IRGENDEINE GARANTIE, sogar ohne die implizite Garantie der MARKTREIFE oder der VERWENDBARKEIT FÜR EINEN BESTIMMTEN ZWECK. Details finden Sie in der Datei GNU General Public License.

CHANGELOG
webmin Release Notes

Webmin 2.021 released March 19, 2023

Add ability to set locale in Webmin Users module for consistency
Fix to preserve initial install directory when upgrading manually
Fix to preserve minimal install type when upgrading manually
Fix an error when make_date is called on undefined value #1860
Fix clearing packages caches before checking for updates in status collection #1863
Update the Authentic theme to the latest version

INFO.sh
source /pkgscripts/include/pkg_util.sh
package="webmin"
displayname="Webmin"
displayname_enu="Webmin"
displayname_ger="Webmin"
version="2.021"
firmware="7.0-40000"
os_min_ver="7.0-40000"
support_center="no"
thirdparty="yes"
reloadui="yes"
arch="noarch"
dsmuidir="ui"
ctl_stop="yes"
startable="yes"
silent_install="no"
silent_upgrade="yes"
silent_uninstall="no"
package_icon="PACKAGE_ICON.PNG"
package_icon_256="PACKAGE_ICON_256.PNG"
dsmappname="SYNO.SDS._ThirdParty.App.webmin"
maintainer="Akay-Devel"
description="Webmin is a web-based interface for system administration. Using any modern web browser, you can setup user accounts, Apache, DNS, file sharing and much more. Webmin removes the need to manually edit Unix configuration files like /etc/passwd, and lets you manage a system from the console or remotely."
description_enu="Webmin is a web-based interface for system administration. Using any modern web browser, you can setup user accounts, Apache, DNS, file sharing and much more. Webmin removes the need to manually edit Unix configuration files like /etc/passwd, and lets you manage a system from the console or remotely."
description_ger="Webmin ist eine Web-basierte Oberfläche für die Systemadministration. Mit jedem modernen Webbrowser können Sie Benutzerkonten, Apache, DNS, File Sharing und vieles mehr einrichten. Mit Webmin entfällt die Notwendigkeit, Unix Konfigurationsdateien wie z.B. /etc/passwd manuell zu bearbeiten und ermöglicht Ihnen die Verwaltung des Systems von der Konsole oder aus der Ferne."
helpurl="https://www.webmin.com"
[ "$(caller)" != "0 NULL" ] && return 0
pkg_dump_info


SPK_Build_Stage.sh
#!/bin/bash
# Filename: SPK_Build_Stage.sh - coded in utf-8

#                webmin
#
#        Copyright (C) 2023 by Akay-Devel 
# Member of the German Synology Community Forum
#             License GNU GPLv3
#   https://www.gnu.org/licenses/gpl-3.0.html


# -------------------------------------------------------------------------------------------------
# Please make your desired changes here
# -------------------------------------------------------------------------------------------------


# Changing the package name
packagename="webmin"

# Changing the copyright
copyright="Copyright (C) 2023 by"

# Changing the maintrainer name
maintrainer="Akay-Devel"

# Change the name of the package directory to [packagename]? yes/no
changedirname="yes" 


# -------------------------------------------------------------------------------------------------
# Please do not make any changes below, unless you know what you are doing!
# -------------------------------------------------------------------------------------------------
old_packagename="webmin"
old_copyright="Copyright (C) 2023 by"
old_maintrainer="Akay-Devel"
path=$(pwd)


# Change file names that contain the package name
# -----------------------------------------------
if [[ "${old_packagename}" != "${packagename}" ]]; then
    mv ./${old_packagename}.c ./${packagename}.c
    chmod 777 ./${packagename}.c
    echo 'File names containing the name [ '${old_packagename}' ] were renamed to [ '${packagename}' ]'
else
    echo 'No file names have been changed'
fi


# Customize Package Name
# -----------------------------------------------
if [[ "${packagename}" != "${old_packagename}" ]]; then
    find . -type f -print0 | xargs -0 -n 1 sed -i -e 's/'"${old_packagename}"'/'"${packagename}"'/g'
    echo 'Package name was changed from [ '${old_packagename}' ] to [ '${packagename}' ]'
else
    echo 'The Package name [ '${packagename}' ] was not changed'
fi


# Costumize Maintrainer
# -----------------------------------------------
if [[ "${maintrainer}" != "${old_maintrainer}" ]]; then
    find . -type f -print0 | xargs -0 -n 1 sed -i -e 's/'"${old_maintrainer}"'/'"${maintrainer}"'/g'
    echo 'Maintrainer was changed from [ '${old_maintrainer}' ] to [ '${maintrainer}' ]'
else
    echo 'Maintrainer [ '${maintrainer}' ] was not changed'
fi


# Costumize Copyright
# -----------------------------------------------
if [[ "${copyright}" != "${old_copyright}" ]]; then
    find . -type f -print0 | xargs -0 -n 1 sed -i -e 's/'"${old_copyright}"'/'"${copyright}"'/g'
    echo 'Copyright was changed from [ '${old_copyright}' ] to [ '${copyright}' ]'
else
    echo 'Copyright [ '${copyright}' ] was not changed'
fi

if [[ "${changedirname}" == "yes" ]]; then
    mv ${path} ${path%/*}/${packagename}
    cd ${path%/*}/${packagename}
    exec bash
else
    exit 0
fi



ui/app_permissions.sh
#!/bin/sh
# Filename: app_permissions.sh - coded in utf-8

#                webmin
#
#        Copyright (C) 2023 by Akay-Devel 
# Member of the German Synology Community Forum
#             License GNU GPLv3
#   https://www.gnu.org/licenses/gpl-3.0.html


# call: /usr/syno/synoman/webman/3rdparty/webmin/app_permissions.sh
app_name="webmin"


# Function: Add or remove users from a group
# --------------------------------------------------------------
# Call: synogroupuser "[adduser or deluser]" "GROUP" "USER"
function synogroupuser()
{
	oldIFS=${IFS}
	IFS=$'\n'
	query=${1}
	group=${2}
	user=${3}
	userlist=$(synogroup --get ${group} | grep -E '^[0-9]*:'| sed -e 's/^[0-9]*:\[\(.*\)\]/\1/')
	updatelist=()
	for i in ${userlist}; do
		if [[ "${query}" == "adduser" ]]; then
			[[ "${i}" != "${user}" ]] && updatelist+=(${i})
			[[ "${i}" == "${user}" ]] && userexists="true"
		elif [[ "${query}" == "deluser" ]]; then
			[[ "${i}" != "${user}" ]] && updatelist+=(${i})
			[[ "${i}" == "${user}" ]] && userexists="true"
		else
			synodsmnotify -c SYNO.SDS._ThirdParty.App.${app_name} @administrators ${app_name}:app:title ${app_name}:app:groupuser_error
			exit 1
		fi
	done
	
	if [[ -z "${userexists}" && "${query}" == "adduser" ]]; then
		updatelist+=(${user})
		synogroup --member ${group} ${updatelist[@]}
		synodsmnotify -c SYNO.SDS._ThirdParty.App.${app_name} @administrators ${app_name}:app:title ${app_name}:app:adduser_true
	elif [[ -n "${userexists}" && "${query}" == "adduser" ]]; then
		synodsmnotify -c SYNO.SDS._ThirdParty.App.${app_name} @administrators ${app_name}:app:title ${app_name}:app:adduser_exists
		exit 2
	elif [[ -n "${userexists}" && "${query}" == "deluser" ]]; then
		synogroup --member ${group} ${updatelist[@]}
		synodsmnotify -c SYNO.SDS._ThirdParty.App.${app_name} @administrators ${app_name}:app:title ${app_name}:app:deluser_true
	elif [[ -z "${userexists}" && "${query}" == "deluser" ]]; then
		synodsmnotify -c SYNO.SDS._ThirdParty.App.${app_name} @administrators ${app_name}:app:title ${app_name}:app:deluser_notexist
		exit 3
	fi
	
	IFS=${oldIFS}
}

# Set App permissions
# --------------------------------------------------------------

	# Check if version corresponds to min. DSM 7
	# ----------------------------------------------------------
	if [ $(synogetkeyvalue /etc.defaults/VERSION majorversion) -ge 7 ]; then
	
		# Add DSM7DemoSPK to the administrators group
		if [[ "${1}" == "adduser" ]]; then
			synogroupuser "adduser" "administrators" "DSM7DemoSPK"
		fi
		
		# Remove DSM7DemoSPK from administrators group
		if [[ "${1}" == "deluser" ]]; then
			synogroupuser "deluser" "administrators" "DSM7DemoSPK"
		fi
		
	fi


ui/modules/parse_url.sh
#!/bin/bash
# Filename: parse_url.sh - coded in utf-8

#                webmin
#
#        Copyright (C) 2023 by Akay-Devel 
# Member of the German Synology Community Forum
#             License GNU GPLv3
#   https://www.gnu.org/licenses/gpl-3.0.html


# urlencode and urldecode https://gist.github.com/cdown/1163649
# --------------------------------------------------------------
function urlencode() {
    # urlencode <string>
    old_lc_collate=$LC_COLLATE
    LC_COLLATE=C

    local length="${#1}"
    for (( i = 0; i < length; i++ )); do
        local c="${1:$i:1}"
        case $c in
            [a-zA-Z0-9.~_-]) printf '%s' "$c" ;;
            *) printf '%%%02X' "'$c" ;;
        esac
    done

    LC_COLLATE=$old_lc_collate
}

function urldecode() {
    # urldecode <string>
    local url_encoded="${1//+/ }"
    printf '%b' "${url_encoded//%/\\x}"
}


ui/modules/parse_language.sh
#!/bin/bash
# Filename: parse_language.sh - coded in utf-8

#                webmin
#
#        Copyright (C) 2023 by Akay-Devel 
# Member of the German Synology Community Forum
#             License GNU GPLv3
#   https://www.gnu.org/licenses/gpl-3.0.html


# Spracheinstellungen konfigurieren
# --------------------------------------------------------------

#********************************************************************#
#  Description: Script get the current used dsm language             #
#  Author:      QTip from the german Synology support forum          #
#  Copyright:   2016-2018 by QTip / Modified 2022 by Akay-Devel          #
#  License:     GNU GPLv3                                            #
#  ----------------------------------------------------------------  #
#  Version:     0.15 - 11/06/2018                                    #
#********************************************************************#

# Declare translation table
declare -A ISO2SYNO
ISO2SYNO=( ["de"]="ger" ["en"]="enu" ["zh"]="chs" ["cs"]="csy" ["jp"]="jpn" ["ko"]="krn" ["da"]="dan" ["fr"]="fre" ["it"]="ita" ["nl"]="nld" ["no"]="nor" ["pl"]="plk" ["ru"]="rus" ["sp"]="spn" ["sv"]="sve" ["hu"]="hun" ["tr"]="trk" ["pt"]="ptg" )

# Standard language
default_lang="enu"

# Script language
script_lang=$(/bin/get_key_value /etc/synoinfo.conf maillang)
if [ "${script_lang}" == "def" ]; then
	script_lang="${default_lang}"
fi

# DSM language
gui_lang=$(/bin/get_key_value /etc/synoinfo.conf language)
if [[ "${gui_lang}" == "def" ]] ; then
	# Determine Browser language
	if [ -n "${HTTP_ACCEPT_LANGUAGE}" ] ; then
		bl=$(echo ${HTTP_ACCEPT_LANGUAGE} | cut -d "," -f1)
		bl=${bl:0:2}
		gui_lang="${ISO2SYNO[${bl}]}"
	else
		gui_lang="${default_lang}"
	fi
fi

# Load language file
if [ -f "lang/lang_${gui_lang}.txt" ]; then
	source "lang/lang_${gui_lang}.txt"
fi


SPK_Pack_Stage.sh
#!/bin/bash
# Filename: SPK_Pack_Stage.sh - coded in utf-8

#                webmin
#
#        Copyright (C) 2023 by Akay-Devel 
# Member of the German Synology Community Forum
#             License GNU GPLv3
#   https://www.gnu.org/licenses/gpl-3.0.html


# -------------------------------------------------------------------------------------------------
# Please make your desired changes here
# -------------------------------------------------------------------------------------------------

# Define package name and version
package_name="webmin"
version="2.021"


# -------------------------------------------------------------------------------------------------
# Please do not make any changes below, unless you know what you are doing!
# -------------------------------------------------------------------------------------------------


# Create working directory
# -----------------------------------------------
path=$(pwd)
working_dir=/tmp/_${package_name}
if [ -d ${working_dir} ]; then
    rm -rf ${working_dir}
else
    mkdir -p ${working_dir}
fi


# Create pack stage
# -----------------------------------------------
package_dir=${working_dir}/package
mkdir -p ${package_dir}
cp -a ui ${package_dir}
cp -a ~/Dokumente/Develop/synology.dev/webmin/working/etc ${package_dir}
cp -a ~/Dokumente/Develop/synology.dev/webmin/working/man ${package_dir}
cp -a ~/Dokumente/Develop/synology.dev/webmin/working/reset ${package_dir}
cp -a ~/Dokumente/Develop/synology.dev/webmin/working/scripts ${package_dir}
cp -a ~/Dokumente/Develop/synology.dev/webmin/working/update ${package_dir}
# cp -a ~/Dokumente/Develop/synology.dev/webmin/working/app ${package_dir}

# Create binary file(s)
# -----------------------------------------------
binary_dir=${package_dir}/usr/local/bin
mkdir -p ${binary_dir}
touch ${binary_dir}/${package_name}
echo "#!/bin/sh" >> ${binary_dir}/${package_name}
echo '/var/packages/'${package_name}'/scripts/start-stop-status "$@"' >> ${binary_dir}/${package_name}


# Create SPK
# -----------------------------------------------
cp -a conf ${working_dir}/conf
cp -a scripts ${working_dir}/scripts
cp -a WIZARD_UIFILES ${working_dir}/WIZARD_UIFILES
cp -a CHANGELOG ${working_dir}
cp -a LICENSE ${working_dir}
cp -a PACKAGE_ICON*.PNG ${working_dir}

sed '1d' INFO.sh > ${working_dir}/INFO.tmp
head -n -2 < ${working_dir}/INFO.tmp > ${working_dir}/INFO
rm ${working_dir}/INFO.tmp

chmod -R 777 ${working_dir}
chmod -R 755 ${package_dir}


# Build SPK
# -----------------------------------------------
tar -C ${package_dir}/ -czf ${working_dir}/package.tgz .
chmod 755 ${working_dir}/package.tgz
rm -rf ${package_dir}
cd ${working_dir}
tar --exclude=".git/*" --exclude=".gitignore/*" -cvf ${working_dir}/${package_name}-${version}.spk *
mv ${working_dir}/${package_name}-${version}.spk ${path}
rm -rf ${working_dir}

